docs/spec/distributing.rst

```diff
@@ -10,72 +10,154 @@ Stub files
 
 (Originally specified in :pep:`484`.)
 
-Stub files are files containing type hints that are only for use by
-the type checker, not at runtime.  There are several use cases for
-stub files:
+*Stub files*, also called *type stubs*, provide type information for untyped
+Python packages and modules. Stub files serve multiple purposes:
+
+* They are the only way to add type information to extension modules.
+* They can provide type information for packages that do not wish to
+  add them inline.
+* They can be distributed separately from the package or module that they
+  provide types for. The latter is referred to as the *implementation*.
+  This allows stubs to be developed at a different pace or by different
+  authors, which is especially useful when adding type annotations to
+  existing packages.
+* They can act as documentation, succinctly explaining the external
+  API of a package, without including implementation details or private
+  members.
+
+Stub files use a subset of the constructs used in Python source files, as
+described in :ref:`stub-file-supported-constructs` below. Type checkers should
+parse a stub that uses only such constructs without error and not interpret any
+construct in a manner contradictory to this specification. However, type
+checkers are not required to implement checks for all of these constructs and
+can elect to ignore unsupported ones. Additionally, type checkers can support
+constructs not described here.
+
+If a stub file is found for a module, the type checker should not read the
+corresponding "real" module. See :ref:`mro` for more information.
+
+.. _stub-file-syntax:
+
+Syntax
+^^^^^^
+
+Stub files are syntactically valid Python files in the earliest Python version
+that is not yet end-of-life. They use a ``.pyi`` suffix. The Python syntax used
+by stub files is independent from the Python versions supported by the
+implementation, and from the Python version the type checker runs under (if
+any).
+
+Type checkers should treat stubs as if ``from __future__ import annotations`` is
+enabled. In particular, forward references do not need to be quoted, and syntax
+from newer versions than otherwise supported may be used in annotation
+expressions. For example, the pipe union syntax (``X | Y``) introduced in Python
+3.10 may be used even before Python 3.9 reaches end-of-life.
+
+.. _stub-file-supported-constructs:
+
+Supported Constructs
+^^^^^^^^^^^^^^^^^^^^
+
+Type checkers should fully support these constructs:
+
+* All features from the ``typing`` module of the latest released Python version
+* Comments, including type declaration (``# type: X``) and error suppression
+  (``type: ignore``) comments
+* Import statements, including the standard :ref:`import-conventions` and cyclic
+  imports
+* Aliases, including type aliases, at both the module and class level
```

Comment:
Should type aliases at the class level be included? I'm not sure they're well-specified and I don't think we use them in typeshed.

Reply:
Type aliases within classes are used in type stubs today. Maybe not in typeshed, but in other libraries. I think it would be problematic not to include them.

Reply:
I believe they're not used in typeshed because pytype doesn't support them in stubs ^^; I don't feel strongly about this either way, but it seems better to allow them unless there's a compelling reason not to.